### ISSUE TITLE
catch macro causes hidden lifetime parameters in types are deprecated 

### ISSUE THREAD

**Issue by valkum** [SergioBenitez/Rocket#1547]
main.rs snippet:
```
#![warn(rust_2018_idioms)]
...
#[catch(404)]
fn not_found_catcher(_req: &'_ Request<'_>) -> String {
    "404 Not Found".to_owned()
}
```
Clippy output:
```
warning: hidden lifetime parameters in types are deprecated
   --> src/main.rs:193:1
    |
193 | fn not_found_catcher(_req: &'_ Request<'_>) -> String {
    | ^^- help: indicate the anonymous lifetime: `<'_>`
    |
note: the lint level is defined here
   --> src/main.rs:1:9
    |
1   | #![warn(rust_2018_idioms)]
    |         ^^^^^^^^^^^^^^^^
    = note: `#[warn(elided_lifetimes_in_paths)]` implied by `#[warn(rust_2018_idioms)]`

```
Macro expansion:
```
fn not_found_catcher(_req: &'_ Request<'_>) -> String {
    "404 Not Found".to_owned()
}
#[doc(hidden)]
#[allow(non_camel_case_types)]
#[doc = r" Rocket code generated proxy structure."]
struct not_found_catcher {
}
#[doc = r" Rocket code generated proxy static conversion implementation."]
impl From<not_found_catcher> for rocket::StaticCatcherInfo {
    fn from(_: not_found_catcher) -> rocket::StaticCatcherInfo {
        fn monomorphized_function<'_b>(__status: rocket::http::Status,
                                       __req: &'_b rocket::request::Request)
         -> rocket::catcher::ErrorHandlerFuture<'_b> {
            ::std::boxed::Box::pin(async move
                                       {
                                           let __response =
                                               {
                                                   let ___responder =
                                                       not_found_catcher(__req);
                                                   ::rocket::response::Responder::respond_to(___responder,
                                                                                             __req)?
                                               };
                                           rocket::response::Response::build().status(__status).merge(__response).ok()
                                       })
        }
        rocket::StaticCatcherInfo{code: ::std::option::Option::Some(404u16),
                                  handler: monomorphized_function,}
    }
}
#[doc = r" Rocket code generated proxy conversion implementation."]
impl From<not_found_catcher> for rocket::Catcher {
    #[inline]
    fn from(_: not_found_catcher) -> rocket::Catcher {
        rocket::StaticCatcherInfo::from(not_found_catcher{}).into()
    }
}
```

Rocket Version: git rev 1f1f44f336e5a172361fc1860461bb03667b1ed2
OS: MacOS

**Comment by jebrosen** [comment]
From a quick glance at the code this probably affects both 0.4 and `master`, and it should be easy to fix by adding `<'_>` in the appropriate places inside the macro.

We currently aren't testing any examples with `#![warn(rust_2018_idioms)]`, but perhaps we should. I vaguely remember encountering similar issues while I was helping port Rocket to Rust 2018, but the `elided_lifetimes_in_paths` lint was never enabled by default and I didn't prioritize it.

**Comment by valkum** [comment]
I pinned it down to https://github.com/SergioBenitez/Rocket/blob/73037c662015c21ac5c32ac5d6de166b8b91edce/core/codegen/src/attribute/catch.rs#L130

Changing that to `#req: &'_b #Request<'_>` fixes the warning. As `#Request` is only a variable I don't know if that would break anything. Is `#Request` replaced with `rocket::request::Request` in any case? If yes, I think that would be the fix and I would submit a PR fixing this. I would also look into other occurrences of `Request` in other macros.

**Comment by jebrosen** [comment]
> If yes, I think that would be the fix and I would submit a PR fixing this.

I for one would be happy to approve and/or merge that PR!

**Comment by jebrosen** [comment]
Fixed by #1549 / e332ee8.
